File: src/UnwindMC/Generation/Ast/Transformations/FixupPointerArithmetics.cs
using System;
using System.Collections.Generic;
using UnwindMC.Util;
using Type = UnwindMC.Analysis.Data.Type;
using static UnwindMC.Generation.Ast.PatternMatching;
using static UnwindMC.Util.PatternMatching;

namespace UnwindMC.Generation.Ast.Transformations
{
    public class FixupPointerArithmetics : NodeTransformerBase
    {
        private readonly IReadOnlyDictionary<string, Type> _variableTypes;

        public FixupPointerArithmetics(IReadOnlyDictionary<string, Type> variableTypes)
        {
            _variableTypes = variableTypes;
        }

        public override BinaryOperatorNode Transform(BinaryOperatorNode node)
        {
            var var = Capture<VarNode>();
            var value = Capture<ValueNode>();
            return Match(node,
                Binary(var % Var(_), C(Operator.Add), value % Value(_)).Then(() => Fixup(node, var, value)),
                Binary(value % Value(_), C(Operator.Add), var % Var(_)).Then(() => Fixup(node, value, var)),
                Binary(var % Var(_), C(Operator.Subtract), value % Value(_)).Then(() => Fixup(node, var, value)),
                Binary(value % Value(_), C(Operator.Subtract), var % Var(_)).Then(() => Fixup(node, value, var)),
                Otherwise(node));
        }

        private BinaryOperatorNode Fixup(BinaryOperatorNode node, VarNode var, ValueNode value)
        {
            var type = _variableTypes[var.Name];
            if (type.IndirectionLevel > 0 || type.IsFunction)
            {
                if (value.Value % type.Size != 0)
                {
                    throw new InvalidOperationException("Value size must be divisible by type size");
                }
                return new BinaryOperatorNode(node.Operator, var, new ValueNode(value.Value / type.Size));
            }
            return node;
        }

        private BinaryOperatorNode Fixup(BinaryOperatorNode node, ValueNode value, VarNode var)
        {
            return Fixup(node, var, value);
        }
    }
}

File: src/UnwindMC/Generation/Ast/Transformations/FixupPointerArithmetics.fs
module FixupPointerArithmetics

open System.Collections.Generic
open Ast
open Type

let transformer (types: IReadOnlyDictionary<string, DataType>): Transformer.T =
    let fixup (op: Operator) (Var name as var) (value: int): Expression =
        let type_ = types.[name]
        match type_ with
        | Pointer _
        | Function ->
            if value % sizeOf type_ <> 0 then
                failwith "Value size must be divisible by type size"
            let newValue = Value (value / sizeOf type_)
            Binary (op, VarRef var, newValue)
        | _ ->
            Binary (op, VarRef var, Value value)

    {
        Transformer.def with
            transformExpression =
                fun t expr ->
                    match expr with
                    | Binary (Operator.Add, VarRef var, Value value) -> fixup Operator.Add var value
                    | Binary (Operator.Add, Value value, VarRef var) -> fixup Operator.Add var value
                    | Binary (Operator.Subtract, VarRef var, Value value) -> fixup Operator.Subtract var value
                    | Binary (Operator.Subtract, Value value, VarRef var) -> fixup Operator.Subtract var value
                    | _ -> Transformer.def.transformExpression t expr
    }
